contribs/bicycle/src/main/java/org/matsim/contrib/bicycle/BicycleTravelTime.java

```diff
@@ -69,66 +63,78 @@ public double getLinkTravelTime(Link link, double time, Person person, Vehicle v
 	 * Negative gradients (downhill):
 	 * Not linear; highest speeds at 5% or 6% gradient; at gradients higher than 6% braking
 	 */
-	private double computeGradientSpeed(Link link, double vehicleLinkSpeed) {
-		
-		double gradientSpeedFactor = 1.;
-		
+	private double computeGradientFactor(Link link) {
+
+		double factor = 1;
 		if (link.getFromNode().getCoord().hasZ() && link.getToNode().getCoord().hasZ()) {
-			Double fromNodeZ = link.getFromNode().getCoord().getZ();
-			Double toNodeZ = link.getToNode().getCoord().getZ();
-			
-			if ((fromNodeZ != null) && (toNodeZ != null)) {
-				if (toNodeZ > fromNodeZ) { // No positive speed increase for downhill, only decrease for uphill
-					gradientSpeedFactor = 1 - 5. * ((toNodeZ - fromNodeZ) / link.getLength()); // 50% reducation at 10% up-slope
-				}
-			}
-			if (gradientSpeedFactor < 0.1) {
-				gradientSpeedFactor = 0.1;
+			double fromZ = link.getFromNode().getCoord().getZ();
+			double toZ = link.getToNode().getCoord().getZ();
+			if (toZ > fromZ) { // No positive speed increase for downhill, only decrease for uphill
+				double reduction = 1 - 5 * ((toZ - fromZ) / link.getLength());
+				factor = Math.max(0.1, reduction); // maximum reduction is 0.1
 			}
 		}
-		
-		return vehicleLinkSpeed * gradientSpeedFactor;
+
+		return factor;
 	}
-	
+
 	// TODO combine this with comfort
-	private double computeSurfaceSpeed(double vehicleLinkSpeed, String surface, String type) {
-		if (type == null || type.equals("cycleway")) {
-			return vehicleLinkSpeed; // Assuming that dedicated cycleways are on good surface like ashalt
+	private double computeSurfaceFactor(Link link) {
+		if (hasNotAttribute(link, BicycleLabels.WAY_TYPE)
```

Comment:
so, I've included this because the old code did too. Do we want the suface factor to be 1.0 if we don't know the waytype? One could imagine that a user only supplies surface attributes but no waytypes